### ISSUE TITLE
is SoftmaxLoss trainable? how it works?

### ISSUE THREAD

**Issue by hjc3613** [UKPLab/sentence-transformers#2872]
I have read the source code about Softmax loss, I have a question about the classifiler:
![image](https://github.com/user-attachments/assets/f079e4ab-0b4f-4593-bb6d-8eaa2288eee9)
the classifier is initialized out of Module, so, when create optimizer, the classifier's weight is not given to AdamOptimizer:
![image](https://github.com/user-attachments/assets/ec5a82ab-1a3f-4fd2-a779-23fa74ebdb4c)

I think the classifier's weight only have gradients but not update when training, Is my idea wrong?

**Comment by ir2718** [comment]
Hi,

I'm not sure what version you're using, but the current version doesn't have a comment at that [line](https://github.com/UKPLab/sentence-transformers/blob/c0fc0e8238f7f48a1e92dc90f6f96c86f69f1e02/docs/sentence_transformer/training_overview.md?plain=1#L593), so the SoftmaxLoss should work as expected.

**Comment by hjc3613** [comment]
> Hi,
> 
> I'm not sure what version you're using, but the current version doesn't have a comment at that [line](https://github.com/UKPLab/sentence-transformers/blob/c0fc0e8238f7f48a1e92dc90f6f96c86f69f1e02/docs/sentence_transformer/training_overview.md?plain=1#L593), so the SoftmaxLoss should work as expected.

pip list | grep sentence-transformers
sentence-transformers         3.0.1

when I debug in the softmaxloss, I found that after some steps, the classifier's weights do not changed at all, my script as bellow:

from datasets import load_dataset
from sentence_transformers import SentenceTransformer, SentenceTransformerTrainer
from sentence_transformers.losses import CoSENTLoss, MultipleNegativesRankingLoss, SoftmaxLoss

all_nli_pair_class_train = load_dataset("sentence-transformers/all-nli", "pair-class", split="train[:1000]",cache_dir=cache_dir)
model = SentenceTransformer("all-MiniLM-L6-v2")
train_dataset = {
    "all-nli-pair-class": all_nli_pair_class_train,
}
softmax_loss = SoftmaxLoss(model, 384, 3)
losses = {
    "all-nli-pair-class": softmax_loss,
}
trainer = SentenceTransformerTrainer(
    model=model,
    train_dataset=train_dataset,
    loss=losses,
)
trainer.train()

my code is followed the document https://sbert.net/docs/sentence_transformer/training_overview.html, only using the softmaxloss:
![image](https://github.com/user-attachments/assets/43343f16-a1d0-4bf7-b042-123af16cd7ec)

**Comment by ir2718** [comment]
@hjc3613 

After some debugging it truly does seem like the SoftmaxLoss.classifiers parameters aren't changing.

Printing out the parameters after every batch yields the following result:
```
  0%|                                                                         | 0/375 [00:00<?, ?it/s]Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

  0%|▏                                                                | 1/375 [00:00<01:39,  3.74it/s]Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

Classifier weights:
Parameter containing:
tensor([[-0.0098, -0.0288, -0.0064,  ...,  0.0199, -0.0013, -0.0175],
        [ 0.0022,  0.0125,  0.0182,  ..., -0.0202, -0.0078,  0.0008],
        [ 0.0075,  0.0273,  0.0131,  ...,  0.0005,  0.0058,  0.0276]],
       device='cuda:0', requires_grad=True) 

  2%|█                                                                | 6/375 [00:00<00:19, 18.72it/s]
```

While printing out the names if the optimizers parameters inside the Trainer.create_optimizer method yields:
```
0.auto_model.embeddings.word_embeddings.weight
0.auto_model.embeddings.position_embeddings.weight
0.auto_model.embeddings.token_type_embeddings.weight
0.auto_model.embeddings.LayerNorm.weight
0.auto_model.embeddings.LayerNorm.bias
0.auto_model.encoder.layer.0.attention.self.query.weight
0.auto_model.encoder.layer.0.attention.self.query.bias
0.auto_model.encoder.layer.0.attention.self.key.weight
0.auto_model.encoder.layer.0.attention.self.key.bias
0.auto_model.encoder.layer.0.attention.self.value.weight
0.auto_model.encoder.layer.0.attention.self.value.bias
0.auto_model.encoder.layer.0.attention.output.dense.weight
0.auto_model.encoder.layer.0.attention.output.dense.bias
0.auto_model.encoder.layer.0.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.0.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.0.intermediate.dense.weight
0.auto_model.encoder.layer.0.intermediate.dense.bias
0.auto_model.encoder.layer.0.output.dense.weight
0.auto_model.encoder.layer.0.output.dense.bias
0.auto_model.encoder.layer.0.output.LayerNorm.weight
0.auto_model.encoder.layer.0.output.LayerNorm.bias
0.auto_model.encoder.layer.1.attention.self.query.weight
0.auto_model.encoder.layer.1.attention.self.query.bias
0.auto_model.encoder.layer.1.attention.self.key.weight
0.auto_model.encoder.layer.1.attention.self.key.bias
0.auto_model.encoder.layer.1.attention.self.value.weight
0.auto_model.encoder.layer.1.attention.self.value.bias
0.auto_model.encoder.layer.1.attention.output.dense.weight
0.auto_model.encoder.layer.1.attention.output.dense.bias
0.auto_model.encoder.layer.1.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.1.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.1.intermediate.dense.weight
0.auto_model.encoder.layer.1.intermediate.dense.bias
0.auto_model.encoder.layer.1.output.dense.weight
0.auto_model.encoder.layer.1.output.dense.bias
0.auto_model.encoder.layer.1.output.LayerNorm.weight
0.auto_model.encoder.layer.1.output.LayerNorm.bias
0.auto_model.encoder.layer.2.attention.self.query.weight
0.auto_model.encoder.layer.2.attention.self.query.bias
0.auto_model.encoder.layer.2.attention.self.key.weight
0.auto_model.encoder.layer.2.attention.self.key.bias
0.auto_model.encoder.layer.2.attention.self.value.weight
0.auto_model.encoder.layer.2.attention.self.value.bias
0.auto_model.encoder.layer.2.attention.output.dense.weight
0.auto_model.encoder.layer.2.attention.output.dense.bias
0.auto_model.encoder.layer.2.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.2.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.2.intermediate.dense.weight
0.auto_model.encoder.layer.2.intermediate.dense.bias
0.auto_model.encoder.layer.2.output.dense.weight
0.auto_model.encoder.layer.2.output.dense.bias
0.auto_model.encoder.layer.2.output.LayerNorm.weight
0.auto_model.encoder.layer.2.output.LayerNorm.bias
0.auto_model.encoder.layer.3.attention.self.query.weight
0.auto_model.encoder.layer.3.attention.self.query.bias
0.auto_model.encoder.layer.3.attention.self.key.weight
0.auto_model.encoder.layer.3.attention.self.key.bias
0.auto_model.encoder.layer.3.attention.self.value.weight
0.auto_model.encoder.layer.3.attention.self.value.bias
0.auto_model.encoder.layer.3.attention.output.dense.weight
0.auto_model.encoder.layer.3.attention.output.dense.bias
0.auto_model.encoder.layer.3.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.3.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.3.intermediate.dense.weight
0.auto_model.encoder.layer.3.intermediate.dense.bias
0.auto_model.encoder.layer.3.output.dense.weight
0.auto_model.encoder.layer.3.output.dense.bias
0.auto_model.encoder.layer.3.output.LayerNorm.weight
0.auto_model.encoder.layer.3.output.LayerNorm.bias
0.auto_model.encoder.layer.4.attention.self.query.weight
0.auto_model.encoder.layer.4.attention.self.query.bias
0.auto_model.encoder.layer.4.attention.self.key.weight
0.auto_model.encoder.layer.4.attention.self.key.bias
0.auto_model.encoder.layer.4.attention.self.value.weight
0.auto_model.encoder.layer.4.attention.self.value.bias
0.auto_model.encoder.layer.4.attention.output.dense.weight
0.auto_model.encoder.layer.4.attention.output.dense.bias
0.auto_model.encoder.layer.4.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.4.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.4.intermediate.dense.weight
0.auto_model.encoder.layer.4.intermediate.dense.bias
0.auto_model.encoder.layer.4.output.dense.weight
0.auto_model.encoder.layer.4.output.dense.bias
0.auto_model.encoder.layer.4.output.LayerNorm.weight
0.auto_model.encoder.layer.4.output.LayerNorm.bias
0.auto_model.encoder.layer.5.attention.self.query.weight
0.auto_model.encoder.layer.5.attention.self.query.bias
0.auto_model.encoder.layer.5.attention.self.key.weight
0.auto_model.encoder.layer.5.attention.self.key.bias
0.auto_model.encoder.layer.5.attention.self.value.weight
0.auto_model.encoder.layer.5.attention.self.value.bias
0.auto_model.encoder.layer.5.attention.output.dense.weight
0.auto_model.encoder.layer.5.attention.output.dense.bias
0.auto_model.encoder.layer.5.attention.output.LayerNorm.weight
0.auto_model.encoder.layer.5.attention.output.LayerNorm.bias
0.auto_model.encoder.layer.5.intermediate.dense.weight
0.auto_model.encoder.layer.5.intermediate.dense.bias
0.auto_model.encoder.layer.5.output.dense.weight
0.auto_model.encoder.layer.5.output.dense.bias
0.auto_model.encoder.layer.5.output.LayerNorm.weight
0.auto_model.encoder.layer.5.output.LayerNorm.bias
0.auto_model.pooler.dense.weight
0.auto_model.pooler.dense.bias
```
This suggests the losses parameters aren't included in the optimization.

This seems like a bug to me but maybe I missed something. @tomaarsen, what do you think?

**Comment by hjc3613** [comment]
the document say that softmaxloss is not the better，may this caused？

**Comment by tomaarsen** [comment]
Hello!

Hmm, I think this is indeed a bug. This classifier is used to convert the sentence embedding back into `num_labels`, e.g. 3 for NLI (e.g. [all-nli](https://huggingface.co/datasets/sentence-transformers/all-nli)), so that we can get a signal to update the sentence embeddings. After training, this classifier is no longer used anywhere - it's not a part of the model.

However, the original `fit` method to train created the optimizer using the loss model, rather than the actual model itself, so the classifier in SoftmaxLoss was indeed included. If I train with SoftmaxLoss in v3.0.1:
```python
from datasets import load_dataset
from sentence_transformers import SentenceTransformer, SentenceTransformerTrainer
from sentence_transformers.evaluation import TripletEvaluator
from sentence_transformers.losses import SoftmaxLoss

all_nli_pair_class_train = load_dataset(
    "sentence-transformers/all-nli",
    "pair-class",
    split="train[:1000]",
)
# Load a dataset with (anchor, positive, negative) triplets
all_nli_pair_class_eval = load_dataset("sentence-transformers/all-nli", "triplet", split="dev")

# Initialize the TripletEvaluator using anchors, positives, and negatives
triplet_evaluator = TripletEvaluator(
    anchors=all_nli_pair_class_eval[:1000]["anchor"],
    positives=all_nli_pair_class_eval[:1000]["positive"],
    negatives=all_nli_pair_class_eval[:1000]["negative"],
    name="all-nli-dev",
)

model = SentenceTransformer("all-MiniLM-L6-v2")
train_dataset = {
    "all-nli-pair-class": all_nli_pair_class_train,
}
softmax_loss = SoftmaxLoss(model, 384, 3)

trainer = SentenceTransformerTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=all_nli_pair_class_eval,
    loss=softmax_loss,
)
trainer.train()
results = triplet_evaluator(model)
print(results)
```
Then I get:
```
{'all-nli-dev_cosine_accuracy': 0.541, 'all-nli-dev_dot_accuracy': 0.459, 'all-nli-dev_manhattan_accuracy': 0.547, 'all-nli-dev_euclidean_accuracy': 0.541, 'all-nli-dev_max_accuracy': 0.547}
```
If I include the classifier in the optimizer, I get:
```
{'all-nli-dev_cosine_accuracy': 0.7, 'all-nli-dev_dot_accuracy': 0.3, 'all-nli-dev_manhattan_accuracy': 0.694, 'all-nli-dev_euclidean_accuracy': 0.7, 'all-nli-dev_max_accuracy': 0.7}
```

I'm not quite sure what the best solution is yet. The optimizer is created based on the `trainer.model` in the `create_optimizer` method that I inherit from the `transformers` `Trainer` in `SentenceTransformerTrainer`. I was hoping that I would not have to override this method, but it seems like perhaps I do have to.

- Tom Aarsen

**Comment by ir2718** [comment]
Are you perhaps accepting PRs for this issue? @tomaarsen

**Comment by tomaarsen** [comment]
@ir2718 I've already prepared a local patch that I think should solve this problem. I'll be making a PR tomorrow morning, but until then, the extra code is this new method that gets added to `SentenceTransformerTrainer` in `sentence_transformers/trainer.py`. If you're interested, you can experiment with it.

Before I finish the PR tomorrow, I'll also add a warning for `SoftmaxLoss` if the transformers version is before 4.43.0, because 4.43.0 is the first version that includes this fix: https://github.com/huggingface/transformers/pull/31875
which allows me to override the `get_optimizer_cls_and_kwargs` to override which model weights get included in the optimizer.

```python
    def get_optimizer_cls_and_kwargs(
        self, args: SentenceTransformerTrainingArguments, model: Optional[SentenceTransformer] = None
    ) -> Tuple[Any, Any]:
        """
        We have to override the optimizer_grouped_parameters because the Trainer superclass bases it on the `model`
        itself, but the SentenceTransformer losses can have weights that should be updated as well, e.g.
        SoftmaxLoss (see #2872).

        This method requires `transformers` >= 4.43.0.
        """

        if isinstance(self.loss, dict):
            loss_model = nn.Sequential(OrderedDict(self.loss))
        else:
            loss_model = self.loss
        optimizer_cls, optimizer_kwargs = super().get_optimizer_cls_and_kwargs(args, loss_model)

        # If the kwargs were not overridden by the super() call, then we should override them here so that the potential
        # weights in the loss(es) can also be updated.
        if not {"params", "model", "optimizer_dict"} & set(optimizer_kwargs.keys()):
            decay_parameters = self.get_decay_parameter_names(loss_model)
            optimizer_kwargs["optimizer_dict"] = [
                {
                    "params": [
                        p for n, p in loss_model.named_parameters() if (n in decay_parameters and p.requires_grad)
                    ],
                    "weight_decay": self.args.weight_decay,
                },
                {
                    "params": [
                        p for n, p in loss_model.named_parameters() if (n not in decay_parameters and p.requires_grad)
                    ],
                    "weight_decay": 0.0,
                },
            ]

        return optimizer_cls, optimizer_kwargs
```

Feel free to let me know if you have any questions about this fix.

- Tom Aarsen
